### PR TITLE
Laboratory work #3, Anastasia Gubarenko - 21FPL2

### PR DESCRIPTION
**Pull request by cybern00dle**

### CODE REVIEW

**Comment by artyomtugaryov** [comment]
Good progress, but speed up to be in time

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -152,7 +166,10 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        for token in encoded_tokens:
+            if token not in self._id2word.keys():
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:170]:
```suggestion
            if token not in self._id2word:
```

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -233,7 +261,18 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        for vertex in vertex1, vertex2:
+            if vertex not in self._matrix[0]:
+                self._matrix[0].append(vertex)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:268]: it is a bad idea to store vertexes in matrix

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 2 above)
Comment (cybern00dle) [reply lab_3_keywords_textrank/main.py]: done, fixed

File: lab_3_keywords_textrank/main.py (hunk 1 above)
Comment (cybern00dle) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -81,7 +90,7 @@ def preprocess_text(self, text: str) -> tuple[str, ...]:
             tuple[str, ...]
                 clean lowercase tokens with no stop-words
         """
-        pass
+        return self._remove_stop_words(self._clean_and_tokenize(text))
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:93]: It is not clear - save the result of `self._clean_and_tokenize(text)` to a variable

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -120,7 +130,11 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
             tokens : tuple[str, ...]
                 sequence of string tokens
         """
-        pass
+        for token in tokens:
+            if token not in self._word2id:
+                self._word2id[token] = 1001 + tokens.index(token)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:135]: can you save 1001 to a variable and use in two places?

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -251,7 +294,10 @@ def is_incidental(self, vertex1: int, vertex2: int) -> int:
                 1 if vertices are incidental, otherwise 0
         If either of vertices is not present in the graph, -1 is returned
         """
-        pass
+        if vertex1 not in self._vertices or vertex2 not in self._vertices:
+            return -1
+        vertex1_index, vertex2_index = self._vertices.index(vertex1), self._vertices.index(vertex2)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:299]: split into two lines

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 5 above)
Comment (cybern00dle) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 3 above)
Comment (cybern00dle) [reply lab_3_keywords_textrank/main.py]: done

**Review by artyomtugaryov — COMMENTED** [review]
Looks good, but fix tests

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**